src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/Cohost/CohostTextDocumentSyncHandler.cs

```diff
@@ -0,0 +1,83 @@
+
+namespace Microsoft.VisualStudio.LanguageServerClient.Razor.Cohost;
+
+[Export(typeof(IRazorCohostTextDocumentSyncHandler)), Shared]
+[method: ImportingConstructor]
+internal class CohostTextDocumentSyncHandler(
+    ProjectSnapshotManagerDispatcher projectSnapshotManagerDispatcher,
+    IRazorProjectService razorProjectService,
+    DocumentSnapshotFactory documentSnapshotFactory,
+    OpenDocumentGenerator openDocumentGenerator,
+    IRazorLoggerFactory loggerFactory) : IRazorCohostTextDocumentSyncHandler
+{
+    private readonly ProjectSnapshotManagerDispatcher _projectSnapshotManagerDispatcher = projectSnapshotManagerDispatcher;
+    private readonly IRazorProjectService _razorProjectService = razorProjectService;
+    private readonly DocumentSnapshotFactory _documentSnapshotFactory = documentSnapshotFactory;
+    private readonly OpenDocumentGenerator _openDocumentGenerator = openDocumentGenerator;
+    private readonly ILogger _logger = loggerFactory.CreateLogger<CohostTextDocumentSyncHandler>();
+
+    public async Task HandleAsync(int version, RazorCohostRequestContext context, CancellationToken cancellationToken)
+    {
+        var textDocument = context.TextDocument.AssumeNotNull();
+        var textDocumentPath = context.TextDocument.FilePath.AssumeNotNull();
+
+        _logger.LogDebug("[Cohost] DidChange for '{document}' with version {version}.", textDocumentPath, version);
+
+        var sourceText = await textDocument.GetTextAsync(cancellationToken).ConfigureAwait(false);
+
+        // TODO: This is to support the non-cohost server and should eventually be removed, and replaced with code that
+        //       directly triggers creation/deletion of the invisible buffers we use for generated files, as necessary,
+        //       rather than using the "project system" because we don't want the project system to actually exist!
+        await _projectSnapshotManagerDispatcher.RunOnDispatcherThreadAsync(() =>
+        {
+            switch (context.Method)
+            {
+                case Methods.TextDocumentDidOpenName:
+                    _razorProjectService.OpenDocument(textDocumentPath, sourceText, version);
+                    break;
+                case Methods.TextDocumentDidChangeName:
+                    _razorProjectService.UpdateDocument(textDocumentPath, sourceText, version);
+                    break;
+                case Methods.TextDocumentDidCloseName:
+                    _razorProjectService.CloseDocument(textDocumentPath);
+                    break;
+                default:
+                    throw new InvalidOperationException("Unsupported method: " + context.Method);
+            }
+        }, cancellationToken).ConfigureAwait(false);
+
+        // To handle multi-targeting, whilst we get given a TextDocument, we have to actually find all of the linked documents
+        // it might have.
+        var documentIds = context.Solution!.GetDocumentIdsWithFilePath(textDocumentPath);
```

Comment:
Consider capturing `context.Solution` in local (potentially with `AssumeNotNull()`. (I'm assuming that `context.Solution` won't change, of course.)